### PR TITLE
OpenAIRE funding external source feature

### LINKED ISSUE #2816: Search/Import Funded Projects from OpenAIRE API to create new Entities
From release plan; assigned to Dimitris at OpenAIRE. 
Estimate: 60 hours
DEPENDS ON "Convert External Source to Entity" FEATURE.  Would require a new external source to support searching the OpenAIRE API

### PR DESCRIPTION
**Pull request by paulo-graca**
## References
* Fixes #2816
* Replaces #3088
* Depends on https://github.com/openaire/openaire-funders-jaxb-model/pull/2

## Description
Search/Import Funded Projects from OpenAIRE API to create new Entities.

## Instructions for Reviewers
This Pull Request replaces #3088 to bring OpenAIRE funding external source to DSpace. It makes use of the OpenAIRE client and integrates the results as imported entities.
**NOTE:** It's dependent on v.2.0 for the OpenAIRE API client - https://github.com/openaire/openaire-funders-jaxb-model/pull/2 and since it isn't yet available, you may need to install it locally for testing.
Because OpenAIRE API doesn't have IDs nor support search by a ID, we are using the URI:
`info:eu-repo/grantAgreement/EC/ICT/231807/EU`
(based on 3.0 guidelines: https://guidelines.openaire.eu/en/latest/literature/field_projectid.html)

List of changes in this PR:
* Adds a specific configuration (both properties and beans)
* Adds a provider and a connector
* It supports paging
* It also supports a token, but for that, you must register at https://services.openaire.eu/uoa-user-management/registeredServices

After deploying your project can try to access to the following link:
https://YOUR-SERVER/server/api/integration/externalsources/openAIREFunding/entries?sort=score,DESC&page=0&size=10&query=epiwork

It is supposed to return json result for that search.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OpenAIREFundingDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+
+        /**
+         * Add metadata oaire.fundingStream
+         * 
+         * @param metadata fundingStream
+         * @return ExternalDataObjectBuilder
+         */
+        public ExternalDataObjectBuilder addFundingStream(String fundingStream) {
+            this.externalDataObject
+                    .addMetadata(new MetadataValueDTO("oaire", "fundingStream", null, null, fundingStream));
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/external/provider/impl/OpenAIREFundingDataProvider.java:397]: It appears this field doesn't exist in out of the box DSpace.  I'm getting this error `bad_dublin_core schema=oaire.fundingStream.null. Metadata field does not exist!`

I believe it's because the `openaire4-types.xml` file isn't listed in our `registry.metadata.load` setting in dspace.cfg.  Could you append it there as part of this PR? https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L808-L820

**Review by tdonohue — CHANGES_REQUESTED** [review]
@paulo-graca : Thanks for this import feature!  I tested it today, and overall it works to query the OpenAIRE API -- I get back results.  However, I cannot import them yet as it looks like the `openaire4-types.xml` metadata registry file isn't enabled in DSpace 7 yet.  It seems like we should enable that by default, as this PR obviously depends on `oaire.*` metadata fields existing in your DSpace.  I noted this inline below as well.

Beyond that, I also reviewed the code & configurations.  Overall, it looks good...but there are a few minor changes/questions I've requested below -- as parts of the code are a bit confusing to me.

File: dspace/config/modules/openaire-client.cfg [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# you can override this settings in your local.cfg file
+openaire.aai.tokenUsage = false
+
+# URL of the OpenAIRE authentication and authorization service 
+openaire.aai.url = https://aai.openaire.eu/oidc/token
+
+# you will be required to register at OpenAIRE (https://services.openaire.eu/uoa-user-management/registeredServices) 
+# and create your service in order to get the following data:
+openaire.aai.clientId = CLIENT_ID_HERE
+openaire.aai.clientSecret = CLIENT_SECRET_HERE
```
[the remark this answers is not in the document]
Comment (paulo-graca) [reply dspace/config/modules/openaire-client.cfg:25]: Yes, OpenAIRE introduced recently a rate limitation for requests (https://develop.openaire.eu/authentication.html). We can still use their REST API but it will be a very limited usage (60 requests per hour). So I provided the two ways to support the API. For testing or sporadic usage we can ignore these settings, but advisable for a production usage. Perhaps I need to have this more clearly.

Regarding properties naming, I followed the principle of the provider in the namespace. The authentication endpoint provider isn't the same as in Rest. I thought that specifying it, those properties could be re-used for other OpenAIRE services that use the same authentication process. But, specifically for this integration, your proposal it's more suitable.

File: dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                    break;
+                default:
+                    // 200 or other
+                    break;
+            }
+
+            // do not close this httpClient
+            result = getResponse.getEntity().getContent();
+        } catch (MalformedURLException e1) {
+            // TODO Auto-generated catch block
```
[the remark this answers is not in the document]
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java:204]: Done

File: dspace-api/src/main/java/org/dspace/external/provider/impl/OpenAIREFundingDataProvider.java (hunk 1 above)
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/external/provider/impl/OpenAIREFundingDataProvider.java:397]: Yes, I added OpenAIRE types to DSpace.cfg. 
Perhaps we never did it because we wanted an optional feature for only for people that want to really use openaire. But as is (where) raises some avoidable setup issues.

File: dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+        HttpClient httpClient = HttpClientBuilder.create().build();
+        HttpResponse getResponse = httpClient.execute(httpPost);
+
+        JSONObject responseObject = null;
+        try (InputStream is = getResponse.getEntity().getContent();
+                BufferedReader streamReader = new BufferedReader(new InputStreamReader(is, "UTF-8"))) {
+            String inputStr;
+            // verify if we have basic json
+            while ((inputStr = streamReader.readLine()) != null && responseObject == null) {
```
[the remark this answers is not in the document]
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java:124]: This feature is based on Orcid external source for persons/authors. 
https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV3AuthorDataProvider.java#L93-L103

I also use JSONObject in other projects, but I think this approach was initially chosen in order to pre-validate data (perhaps there can be some issues with Orcid and responses). I'm not passionate in any of them, I can change it if you prefer the other way.

File: dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            throw new IOException("Cannot grab OpenAIRE token with nulls service url, client id or secret");
+        }
+
+        String auth = clientId + ":" + clientSecret;
+        String encodedAuth = Base64.getEncoder().encodeToString(auth.getBytes());
+        String authHeader = "Basic " + new String(encodedAuth);
+
+        HttpPost httpPost = new HttpPost(tokenServiceUrl);
+        httpPost.addHeader("Accept", "application/json");
+        httpPost.addHeader("User-Agent", "DSpace/7.X");
```
[the remark this answers is not in the document]
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java]: Done.

File: dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/external/OpenAIRERestConnector.java:125]: Thanks for pointing that out. This is fine as-is then...I'm fine with it since we've the same sort of code elsewhere.

**Review by tdonohue — COMMENTED** [review]
@paulo-graca : I re-tested this today & it's working well now that all the "oaire" metadata fields are created.

One final request here... since you've setup your code similar to the ORCID provider, I wonder if you could also create some *tests* similar to `OrcidExternalSourcesIT`: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/OrcidExternalSourcesIT.java  (This one loads some mock responses from https://github.com/DSpace/DSpace/tree/main/dspace-api/src/test/resources/org/dspace/authority/orcid)

Ideally, we should be providing some basic tests for all new code, including External Source Providers.  So, either this should have some basic ITs like ORCID, or you can add some "Unit Tests" similar to Sherpa: https://github.com/DSpace/DSpace/tree/main/dspace-api/src/test/java/org/dspace/app/sherpa   (This one uses mock responses from https://github.com/DSpace/DSpace/tree/main/dspace-api/src/test/resources/org/dspace/app/sherpa)

Either approach is fine.

Beyond that, I approve of this PR.  It's working for me at least at a basic level.  Thanks!

**Comment by paulo-graca** [comment]
Hi @tdonohue ! I've just added Integration and Unit tests for this feature. They will use provided XML files to simulate API requests.

**Review by tdonohue — APPROVED** [review]
👍 @paulo-graca : This is looking good now. Thanks for all the tests!  I retested and it's still working well for me.  I do have a few very minor requests below (just realignment to make demo responses easier to read), but I'll go ahead and approve it.

@abollini : You'll obviously want to look at this also with regards to the upcoming work to allow entity types to be "mapped" to external providers (DSpace/RestContract#169).  We may want to merge this PR earlier so that you can add a mapping for this provider to OrgUnit to support the [OpenAIRE entities model](https://github.com/DSpace/DSpace/blob/main/dspace/config/entities/openaire4-relationships.xml#L77).

File: dspace-server-webapp/src/test/resources/org/dspace/external/provider/impl/openaire-no-projects.xml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+			   <page>1</page>
+			   <total>0</total>
+							<fields>
+							</fields>
+			</header>
+	  <results>
+									</results>
+	  <browseResults>
+								</browseResults>
+</response>
```
[the remark this answers is not in the document]
Comment (paulo-graca) [reply dspace-server-webapp/src/test/resources/org/dspace/external/provider/impl/openaire-no-projects.xml:16]: @tdonohue thank you Tim for your review and feedback. I've just fixed the XML indentation. Didn't noticed sorry. But it's fixed now.

**Review by tdonohue — APPROVED** [review]
👍 Thanks for the minor xml alignment fixes, @paulo-graca !  This all looks good to me now.

**Comment by tdonohue** [comment]
@paulo-graca : I guess I spoke too soon...your cleanup of the alignment strangely broke your tests?  I'm not sure why but looks like there are now test failures in your new tests.

**Comment by paulo-graca** [comment]
@tdonohue I re did the XML indentation, now, I'm not using eclipse for it. The result passed all my local tests. And besides some break lines I didn't noticed any evident changes. Let's wait for the result.

**Review by tdonohue — APPROVED** [review]
👍 Looks good now, @paulo-graca !  I've verified all the recent changes (since I last tested this) are all simply XML realignment of test files...and those now succeed in GitHub CI.  Thanks again!

**Review by abollini — COMMENTED** [review]
I have reviewed the code and it looks good enough , there are just a couple of minor things as inline comments that I would ask to fix if we decide to accept this implementation.

This same feature is currently available in DSpace-CRIS https://github.com/4Science/DSpace/blob/dspace-cris-7/dspace/config/spring/api/external-services.xml#L203 and was planned to be contributed to dspace as part of the https://github.com/DSpace/DSpace/pull/3143 
The implementation here differ from the DSpace-CRIS implementation as this one use JAXB to parse the OpenAIRE response and their provided model library, instead DSpace-CRIS uses the Live Import Framework.
We are not fun of the Live Import Framework but it was decided to rely on it when the BTE framework was removed. The JAXB approach is much more readable but doesn't allow to configure the mapping between the openaire metadata and the dspace metadata.
We already have place where a JAXB or a more manual parsing approach is in place so this should be not the final argument for a decision, and again using JAXB have some advantage... for instance we suggest to rely on JAXB / external provided data model library when the structure is complex and well maintained by the third party (i.e. ORCID).

In short I vote +0 on that and I would be ok in merging once solved the minor code comment if other agree to don't use the live import for this specific external provider

**Review by tdonohue — APPROVED** [review]
@abollini and @paulo-graca :  I reviewed this PR again today, per @abollini 's comments.  In my opinion, it's OK as-is & we should accept it in 7.1

Here's why I feel it's OK:
1. It appears to me, that the usage of JAXB in this PR was _necessitated_ by the usage of the `eu.openaire.funders-model` dependency provided by OpenAIRE.  Notice that we use the `eu.openaire.jaxb.helper.OpenAIREHandler` and `eu.openaire.jaxb.model.Response`
2. As stated by @abollini , this is similar to why we also use JAXB to parse ORCID... similarly, we are using ORCID provided dependencies to do the main parsing of the response (via JAXB), e.g. `org.orcid.jaxb.model.v3.release.common.OrcidIdentifier`.
3. Therefore, I feel this OpenAIRE scenario is quite similar to the existing ORCID scenario. We're using external APIs to simplify our maintenance, and those APIs both already use JAXB.

That said, as we look at #3143 (perhaps we can bring this forward for 7.2), I'm OK with us looking at ways to enhance the code in this PR to make it more configurable -- that is one of the main reasons for the "Live Import Framework", as field mappings are MUCH easier to configure (in a Spring config file).  In the JAXB approach, those field mappings are in Java code, so they are harder to customize.

In summary, I think we should include this in 7.1.  However, I welcome more discussion in 7.2, especially as we look more at the work of #3143. 

@abollini or @paulo-graca : Any final thoughts, comments or questions about this approach?  I'll wait to merge this PR until next week (end of day Monday at the earliest) to allow you both to add final comments as needed.
